File: VB/Unbound Columns/ViewModels/MainViewModel.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports DevExpress.Data
Imports DevExpress.Mvvm.CodeGenerators
Imports DevExpress.Mvvm.Native
Imports DevExpress.Mvvm.Xpf

Namespace Unbound_Columns.ViewModels

    <GenerateViewModel>
    Public Partial Class MainViewModel

        <GenerateProperty>
        Private _columns As ObservableCollection(Of Column)

        <GenerateProperty>
        Private _items As ObservableCollection(Of Item)

        Public Sub New()
            _items = New ObservableCollection(Of Item)(Enumerable.Range(0, 10).[Select](Function(i) New Item With {.Id = i}))
            _items.ForEach(Sub(x)
                x("Name") = $"Item {_items.IndexOf(x)}"
                x("CreatedAt") = Date.Now.AddDays(_items.IndexOf(x))
            End Sub)
            _columns = New ObservableCollection(Of Column) From {New(_, _) With {.FieldName = "Id"}, New(_, _) With {.FieldName = "Name", .UnboundType = UnboundColumnType.String}, New(_, _) With {.FieldName = "CreatedAt", .UnboundType = UnboundColumnType.DateTime}}
        End Sub

        <GenerateCommand>
        Public Sub OnCustomUnboundColumnData(ByVal e As UnboundColumnRowArgs)
            If e.IsGetData Then e.Value = _items(e.SourceIndex)(e.FieldName)
            If e.IsSetData Then _items(e.SourceIndex)(e.FieldName) = e.Value
        End Sub

        <GenerateCommand>
        Public Sub AddColumn()
            _columns.Add(New Column With {.FieldName = $"Value {_columns.Count}", .UnboundType = UnboundColumnType.Integer})
        End Sub
    End Class

    <GenerateViewModel>
    Public Partial Class Item
        Implements INotifyPropertyChanged

        Private ReadOnly _customFieldValues As Dictionary(Of String, Object) = New(_, _)()

        <GenerateProperty>
        Private _id As Integer

        Default Public Property Item(ByVal fieldName As String) As Object
            Get
                Dim value As Object = Nothing
                _customFieldValues.TryGetValue(fieldName, value)
                Return value
            End Get

            Set(ByVal value As Object)
                _customFieldValues(fieldName) = value
                OnPropertyChanged(fieldName)
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Overridable Sub OnPropertyChanged(<CallerMemberName> ByVal Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class

    <GenerateViewModel>
    Public Partial Class Column

        <GenerateProperty>
        Private _fieldName As String

        <GenerateProperty>
        Private _unboundType As UnboundColumnType?
    End Class
End Namespace

File: CS/Unbound Columns/ViewModels/MainViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Windows.Input;
using DevExpress.Data;
using DevExpress.Mvvm;
using DevExpress.Mvvm.Native;
using DevExpress.Mvvm.Xpf;

namespace Unbound_Columns.ViewModels {
    public class MainViewModel : ViewModelBase {
        public ObservableCollection<Column> Columns { get => GetProperty(() => Columns); set => SetProperty(() => Columns, value); }
        public ObservableCollection<Item> Items { get => GetProperty(() => Items); set => SetProperty(() => Items, value); }

        public MainViewModel() {
            Items = new ObservableCollection<Item>(Enumerable.Range(0, 10).Select(i => new Item { Id = i }));
            Items.ForEach(x => {
                x["Name"] = $"Item {Items.IndexOf(x)}";
                x["CreatedAt"] = DateTime.Now.AddDays(Items.IndexOf(x));
            });

            Columns = new ObservableCollection<Column> {
                new() { FieldName = "Id" },
                new() { FieldName = "Name", UnboundType = UnboundColumnType.String },
                new() { FieldName = "CreatedAt", UnboundType = UnboundColumnType.DateTime }
            };

            OnCustomUnboundColumnDataCommand = new DelegateCommand<UnboundColumnRowArgs>(OnCustomUnboundColumnData);
            AddColumnCommand = new DelegateCommand(AddColumn);
        }

        public void OnCustomUnboundColumnData(UnboundColumnRowArgs e) {
            if (e.IsGetData) e.Value = Items[e.SourceIndex][e.FieldName];

            if (e.IsSetData) Items[e.SourceIndex][e.FieldName] = e.Value;
        }

        public void AddColumn() {
            Columns.Add(new Column { FieldName = $"Value {Columns.Count}", UnboundType = UnboundColumnType.Integer });
        }

        public ICommand OnCustomUnboundColumnDataCommand { get; }
        public ICommand AddColumnCommand { get; }
    }

    public class Item : BindableBase {
        private readonly Dictionary<string, object> _customFieldValues = new();
        public int Id { get => GetProperty(() => Id); set => SetProperty(() => Id, value); }

        public object this[string fieldName] {
            get {
                _customFieldValues.TryGetValue(fieldName, out var value);
                return value;
            }

            set {
                _customFieldValues[fieldName] = value;
                RaisePropertyChanged(fieldName);
            }
        }
    }

    public class Column : BindableBase {
        public string FieldName { get => GetProperty(() => FieldName); set => SetProperty(() => FieldName, value); }
        public UnboundColumnType? UnboundType { get => GetProperty(() => UnboundType); set => SetProperty(() => UnboundType, value); }
    }
}